lib/Sema/AsyncCallerExecutionMigration.cpp

```diff
@@ -0,0 +1,178 @@
+//===-- Sema/AsyncCallerExecutionMigration.cpp ------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file implements code migration support for the `AsyncCallerExecution`
+/// feature.
+///
+//===----------------------------------------------------------------------===//
+
+#include "AsyncCallerExecutionMigration.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/DiagnosticsSema.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/TypeRepr.h"
+#include "swift/Basic/Assertions.h"
+#include "swift/Basic/Feature.h"
+#include "swift/Basic/TaggedUnion.h"
+#include "llvm/ADT/PointerUnion.h"
+
+using namespace swift;
+
+namespace {
+class AsyncCallerExecutionMigrationTarget {
+  ASTContext &ctx;
+  PointerUnion<ValueDecl *, AbstractClosureExpr *, FunctionTypeRepr *> node;
+  TaggedUnion<ActorIsolation, FunctionTypeIsolation> isolation;
+
+public:
+  AsyncCallerExecutionMigrationTarget(ASTContext &ctx, ValueDecl *decl,
+                                      ActorIsolation isolation)
+      : ctx(ctx), node(decl), isolation(isolation) {}
+
+  AsyncCallerExecutionMigrationTarget(ASTContext &ctx,
+                                      AbstractClosureExpr *closure,
+                                      ActorIsolation isolation)
+      : ctx(ctx), node(closure), isolation(isolation) {}
+
+  AsyncCallerExecutionMigrationTarget(ASTContext &ctx, FunctionTypeRepr *repr,
+                                      FunctionTypeIsolation isolation)
+      : ctx(ctx), node(repr), isolation(isolation) {}
+
+  /// Warns that the behavior of nonisolated async functions will change under
+  /// `AsyncCallerExecution` and suggests `@execution(concurrent)` to preserve
+  /// the current behavior.
+  void diagnose() const;
+};
+} // end anonymous namespace
+
+void AsyncCallerExecutionMigrationTarget::diagnose() const {
+  const auto feature = Feature::AsyncCallerExecution;
+
+  ASSERT(node);
+  ASSERT(ctx.LangOpts.getFeatureState(feature).isEnabledForAdoption());
+
+  AbstractFunctionDecl *functionDecl = nullptr;
+  ClosureExpr *closure = nullptr;
+  FunctionTypeRepr *functionRepr = nullptr;
+
+  if (auto *decl = node.dyn_cast<ValueDecl *>()) {
+    // Diagnose only explicit nodes.
+    if (decl->isImplicit()) {
+      return;
+    }
+
+    // Diagnose only functions.
+    functionDecl = dyn_cast<AbstractFunctionDecl>(decl);
```

Comment:
Nit: It's allowed only on `FuncDecl`s at the moment.

Reply:
https://github.com/swiftlang/swift/pull/80032

Reply:
Fantastic, thank you!